### PR TITLE
support the CONNECT method

### LINKED ISSUE #2748: Support CONNECT method
CONNECT method has no :path. 

The current implementation requires a non-empty path somehow:

https://github.com/lucas-clemente/quic-go/blob/bed802aee5732c8fe373bf4f71e789bea2f7f11f/http3/request.go#L40-L47

Also url.ParseRequestURI will return an error on empty path.

### PR DESCRIPTION
**Pull request by klzgrad**
Fix #2748.

### CODE REVIEW

**Comment by klzgrad** [comment]
@googlebot I signed it!

**Comment by klzgrad** [comment]
I don't understand this warning, "string `CONNECT` has 3 occurrences, make it a constant (goconst)." Isn't there only one occurrence?

File: http3/request.go [hunk 1]
```diff
@@ -37,13 +37,28 @@ func requestFromHeaders(headers []qpack.HeaderField) (*http.Request, error) {
 		httpHeaders.Set("Cookie", strings.Join(httpHeaders["Cookie"], "; "))
 	}
 
-	if len(path) == 0 || len(authority) == 0 || len(method) == 0 {
+	isConnect := method == "CONNECT"
```
Comment (marten-seemann) [inline http3/request.go:40]: Can you use `http.MethodConnect` instead?

File: http3/request.go [hunk 2]
```diff
@@ -37,13 +37,28 @@ func requestFromHeaders(headers []qpack.HeaderField) (*http.Request, error) {
...
+			return nil, errors.New(":path must be empty and :authority must not be empty")
+		}
+	} else if len(path) == 0 || len(authority) == 0 || len(method) == 0 {
 		return nil, errors.New(":path, :authority and :method must not be empty")
 	}
 
-	u, err := url.ParseRequestURI(path)
-	if err != nil {
-		return nil, err
+	var url_ *url.URL
```
Comment (marten-seemann) [inline http3/request.go:49]: Nit: Not a fan of a variable name followed by a _. Can you leave this at `u`?

File: http3/request_test.go [hunk 3]
```diff
@@ -81,6 +81,16 @@ var _ = Describe("Request", func() {
...
 	})
 
+	It("handles CONNECT method", func() {
+		headers := []qpack.HeaderField{
+			{Name: ":authority", Value: "quic.clemente.io"},
+			{Name: ":method", Value: "CONNECT"},
+		}
+		req, err := requestFromHeaders(headers)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(req.Method).To(Equal("CONNECT"))
```
Comment (marten-seemann) [inline http3/request_test.go:91]: Better: `http.MethodConnect`.

File: http3/request_test.go [hunk 4]
```diff
@@ -81,6 +81,16 @@ var _ = Describe("Request", func() {
 		}))
 	})
 
+	It("handles CONNECT method", func() {
+		headers := []qpack.HeaderField{
+			{Name: ":authority", Value: "quic.clemente.io"},
+			{Name: ":method", Value: "CONNECT"},
+		}
+		req, err := requestFromHeaders(headers)
```
Comment (marten-seemann) [inline http3/request_test.go:89]: Should we test for the `req.RequestURI` field?

**Review by marten-seemann — COMMENTED** [review]
Thank you @klzgrad! This looks very good to me, I just found two very minor nits (see comments).

I also have a few questions from reading the spec:
> * The ":authority" pseudo-header field contains the host and port to connect to (equivalent to the authority-form of the request-target of CONNECT requests; see Section 3.2.3 of [HTTP11])

We're currently just copying over the `:authority` field. Am I right to assume that the upstream proxy implementation will perform a check that this field actually contains a host and port?

Furthermore, I'm wondering about this paragraph:
> Once the CONNECT method has completed, only DATA frames are permitted to be sent on the stream. Extension frames MAY be used if specifically permitted by the definition of the extension. Receipt of any other frame type MUST be treated as a connection error of type H3_FRAME_UNEXPECTED.

Do we need to check for this?

**Comment by klzgrad** [comment]
> Am I right to assume that the upstream proxy implementation will perform a check that this field actually contains a host and port?

caddyserver/forwardproxy does this. In abstract, we are passing a host and port in a single authority string to the upstream proxy, and to actually perform the connection the proxy always has to first parse the string into a host and port. Checking its validity seems more suitable in that place.

> only DATA frames are permitted to be sent on the stream

Currently checking of this part is done in this code,

https://github.com/lucas-clemente/quic-go/blob/c5a132f158532ee8fbc0442f7c258c929fd77aab/http3/body.go#L50-L72

Not strictly conforming, as it merely ignores HEADERS frame, but the non-conformance is on the same level with regular HTTP requests. As in https://tools.ietf.org/id/draft-ietf-quic-http-29.html#section-4.1-6:

> Receipt of an invalid sequence of frames MUST be treated as a connection error of type H3_FRAME_UNEXPECTED (Section 8). In particular, a DATA frame before any HEADERS frame, or a HEADERS or DATA frame after the trailing HEADERS frame is considered invalid.

I hope fixing all that won't be required in this PR.

**Comment by marten-seemann** [comment]
That makes sense. We can fix that in a separate issue.

**Review by marten-seemann — APPROVED** [review]
LGTM now. Thank you @klzgrad!

**Pull request merged**